####impute with IMPUTE2

##wrap function to impute with IMPUTE2 for batchtools
#data as defined by batchtools call, which is the data to be imputed
#job and instance is defined by batchtools call
#exec expexts the path to the IMPUTE2 Executable
#outpref expects a path with a prefix of how the output files should be named.
#     the full name will be generated by the function including the phasing process, imputation process and chunk
#map.no.chr expects the path to the map file without a chromosome column
#reference.hap expects the path to the reference panel in hap format
# int marks the start of the chunk
#chr.flag needs to be set according to which chromosome is the imputation carried out on

impute2_wrap <- function(data,
                         job,
                         instance,
                         exec,
                         outpref,
                         map.no.chr,
                         reference.hap,
                         int,
                         chr.flag) {
   
    #check cromosome flag for right chunking
    if(chr.flag==19){
        inte <- int + chunk_19 - 1
    }
    if(chr.flag==22){
        inte <- int + chunk_22 - 1
    }
    
    
    #define chunks
    intstring1 <- toString(format(int, scientific = F))
    intstring2 <- toString(format(inte, scientific = F))
    
    #impute with IMPUTE2 in external function
    impdata <- impute2_impute(instance[1],
                              exec,
                              paste(outpref, instance[2], sep = "_"),
                              map.no.chr,
                              reference.hap, 
                              intstring1,
                              intstring2,
                              chr.flag)
    
    # convert imputed data into vcf format
    impute_conv2(instance[1], impdata)
}

#function to impute with IMPUTE2
impute2_impute <- function(data,
                           exec, 
                           outpref,
                           map,
                           reference.hap,
                           int1,
                           int2,
                           chr.flag) {
    
    
    #data is a prefix, reference.haps is a prefix, right sufix will be picked by funcion automatically
    out <- paste(outpref, "_", chr.flag, "_", int1, sep = "")
    
    #set seed
    draw_seed <- sample(1:9999, 1)
    
    #check, which phasing was used because of different formats
    if (dirname(data) == dir_out_shapeit) {
        haplotypes <- paste(data, ".haps", sep = "")
    } else {
        haplotypes <- paste(data, ".haps.gz", sep = "")
    }
    
    
    #execute imputation with IMPUTE2
    system2(exec, 
            c("-use_prephased_g",
              "-m", map,
              "-h", paste(reference.hap, "hap", "gz", sep = "."), 
              "-l", paste(reference.hap, "legend", "gz", sep = "."),
              "-known_haps_g", haplotypes,
              "-int", int1, int2,
              "-seed", draw_seed,
              "-o", paste(out, ".gen", sep = ""),
              "-o_gz")) 
    #return imputed file
    return(out)
}

#function to convert imputed data set into vcf
# insample is path to .sample file, ingen is the path of the imputed data set by IMPUTE2. both without suffix
impute_conv2 <- function(insample, ingen) {
    #takes original . sample file and moves it to the imputed file
    file.copy(paste(insample, ".sample", sep = ""), dirname(ingen))
  #rename .sample file to ensure compatibility
    file.rename(file.path(dirname(ingen), paste(basename(insample), ".sample", sep = "")),
                file.path(dirname(ingen), paste(basename(ingen), ".samples", sep = "")))
    
    #convert to vcf with bcftools
    system2("bcftools", c("convert -o", paste(ingen, ".vcf.gz", sep = ""), "-Oz -G", ingen))
  
    #return converted file path
    return(paste(ingen, ".vcf.gz", sep = ""))
}
