#####imputation with Impute4



##wrap function to impute with IMPUTE4 for batchtools
#data as defined by batchtools call, which is the data to be imputed
#job and instance is defined by batchtools call
#exec expexts the path to the IMPUTE4 Executable
#outpref expects a path with a prefix of how the output files should be named.
#     the full name will be generated by the function including the phasing process, imputation process and chunk
#map.no.chr expects the path to the map file without a chromosome column
#reference.hap expects the path to the reference panel in hap format
# int marks the start of the chunk
#chr.flag needs to be set according to which chromosome is the imputation carried out on

impute4_wrap <- function(data,
                         job,
                         instance,
                         exec,
                         outpref,
                         map.no.chr,
                         reference.hap,
                         int,
                         chr.flag) {
    
    #check cromosome flag for right chunking
    if(chr.flag==19){
        inte <- int + chunk_19 - 1
    }
    if(chr.flag==22){
        inte <- int + chunk_22 - 1
    }
    
    
    #define chunks
    intstring1 <- toString(format(int, scientific = F))
    intstring2 <- toString(format(inte, scientific = F))
    
   
    
    #impute in external function
    impdata <- impute4_impute(instance[1],
                              exec,
                              paste(outpref, instance[2], sep = "_"),
                              map.no.chr,
                              reference.hap, 
                              intstring1, 
                              intstring2,
                              chr.flag)
    
    # convert imputed data into vcf format
    impute_conv2(instance[1], impdata)
    
}

#impute with IMPUTE4
impute4_impute <- function(data, 
                           exec, 
                           outpref,
                           map,
                           reference.hap,
                           int1,
                           int2,
                           chr.flag) {
    
  #construct output name
    out <- paste(outpref, chr.flag, int1, sep = "_")
  #set seed
      draw_seed <- sample(1:9999, 1)
    
      #check, which phasing was used because of different formats
    if (dirname(data) == dir_out_shapeit) {
        haplotypes <- paste(data, ".haps", sep = "")
    } else {
        haplotypes <- paste(data, ".haps.gz", sep = "")
    }
    #execute imputation with IMPUTE4
    system2(exec, c("-m", map, "-h", paste(reference.hap, "hap", "gz", sep = "."),
                    "-l", paste(reference.hap, "legend", "gz", sep = "."),
                    "-g", haplotypes, 
                    "-int", int1, int2,
                    "-seed", draw_seed, "-o", out, "-o_gz", "-maf_align"))  
    
    return(out)
}
