#Impute with PBWT and preparation functions


##wrap function to impute with PBWT for batchtools
#data as defined by batchtools call, which is the data to be imputed
#job and instance is defined by batchtools call
#exec expexts the path to the PBWT executable
#outpref expects a path with a prefix of how the output files should be named.
#     the full name will be generated by the function including the phasing process, imputation process and chunk
#reference.pbwt is the path to the reference panel in pbwt format, but the suffix .pbwt must be omitted
#chr.flag needs to be set according to which chromosome is the imputation carried out on

pbwt_wrap <- function(data,
                      job, 
                      instance,
                      exec,
                      outpref,
                      reference.pbwt,
                      chr.flag){


  
  #impute with PBWT
  system2(exec, c("-readVcfGT", 
                  paste(instance[1], "_pbwt.vcf.gz", sep=""), 
                  "-referenceImpute", file_path_sans_ext(reference.pbwt), 
                  "-imputeMissing -writeVcfGz", paste(outpref,"_",chr.flag,"_", instance[2], ".vcf.gz",sep="")
                  
                  )
          )
  
 #return imputed file
  return(paste(outpref, "_", chr.flag,"_", instance[2], ".vcf.gz",sep=""))
}



#preparing phased data for imputation with pbwt
#this function is called by phasing wrap functions
#data is without suffix
#reference is 
pbwt_prep <- function(data, ref, phasing.flag, chr.flag){

  #check which chromosome is used
  if(chr.flag==22){
    samples_pbwt <- dir_input_samples_22
  }
  
  if(chr.flag==19){
    samples_pbwt <- dir_input_samples_19
  }

  #index file 
  system2(dir_exec_bcftools, c("index -f", paste(data,".vcf.gz",sep="")))
 
  
  #file from phasing, which will be merged with SNPs from reference panel, to indicate which SNPs are missing
  no_format <- file.path(dirname(data),paste("format_fields_removed",phasing.flag,".bcf.gz", sep=""))
  
  #name of file, which is the merged product
  mergename<-paste("merged",chr.flag, phasing.flag,".bcf.gz", sep="")
 
  #name of final output file, which is the test file with added missing SNPs from the reference panel 
  new_data_pbwt<-paste(data,"_pbwt.vcf.gz",sep="")


  #annotate phased file, delete obsolete columns
  system2(dir_exec_bcftools, c("annotate -x INFO,FORMAT,^FORMAT/GT", "-Ob -o", no_format, paste(data,".vcf.gz",sep="")))
  
  #index file
  system2(dir_exec_bcftools, c("index -f", no_format))
  
  #merge with reference panel
  system2(dir_exec_bcftools,c("merge -Ob -o", file.path(dirname(data),mergename),
                              no_format, ref))
  
  #index merged file
  system2(dir_exec_bcftools, c("index -f",file.path(dirname(data),mergename)))
  
  #extract individuals from the original test samples
  #this is the file which will be imputed in the next step
  system2(dir_exec_bcftools, c("view -S", samples_pbwt,
                               "-Oz -o", new_data_pbwt, 
                       file.path(dirname(data),mergename)))
  
  #index new file
  system2(dir_exec_bcftools, c("index -f", new_data_pbwt))
  
  
  #remove obsolete files
  file.remove(file.path(dirname(data),mergename))
  file.remove(no_format)
  file.remove(paste(file.path(dirname(data),mergename),".csi",sep=""))
  
  #return path to prepared pbwt files
  return(new_data_pbwt)
  
}


#extract samples which should be imputed from reference panel
#this funtion ist used in preparation of batchtools
#in.file points towards the reference panel in vcf format
#out.file points to the path where the list should be saved at
pbwt_get_samples<-function(in.file, out.file ){
  
  
  system2(dir_exec_bcftools, c("query -l", in.file, ">", out.file))
  return(out.file)
}

